<!--BOOK_INFORMATION-->
<img align="left" style="padding-right:10px;" src="./figures/cover-small.jpg">

*Este libro es una versión al español de [Python for Everybody](https://www.py4e.com/) escrito por el [Dr. Charles R. Severance](http://www.dr-chuck.com/); este contenido esta disponible en [GitHub](https://github.com/csev/py4e).*

Detalles de Copyright

*Copyright ~ 2009- Charles Severance.
Este trabajo está registrado bajo una Licencia Creative Commons AttributionNonCommercial-ShareAlike 3.0 [CC BY-NC-SA](https://creativecommons.org/licenses/by-nc-sa/3.0/).*

<!--NAVIGATION-->
| [Indice](indice.ipynb) | 

< [Capítulo 1 - ¿Por qué deberías aprender a escribir código?](cap01.ipynb) | [Capítulo 3 - Ejecuciones condicionales](cap03.ipynb) >

# Capítulo 2 - Variables, expresiones y declaraciones

## Valores y tipos

Un valor es una de las cosas básicas con las que trabaja un programa, como una letra o un número. Los valores que hemos visto hasta ahora son 1, 2 y "Hola mundo!".

Estos valores pertenecen a diferentes tipos: 2 es un número entero, y "¡Hola, mundo!" es una cadena, llamada así porque contiene una "cadena" de letras. Usted (y el intérprete) pueden identificar cadenas porque están entre comillas.

La sentencia *print* también funciona para enteros.

In [1]:
print(4)

4


Si no está seguro de qué tipo tiene un valor, el intérprete puede decírselo

In [2]:
type('Hola mundo!')

str

In [3]:
type(17)

int

Como era de esperar, las cadenas pertenecen al tipo *str* y los enteros pertenecen al tipo *int*. Menos obviamente, los números con un punto decimal pertenecen a un tipo llamado *float*, porque estos números se representan en un formato llamado punto flotante.

In [4]:
type(3.2)

float

¿Qué pasa con valores como "17" y "3.2"? Se parecen a números, pero están entre comillas como cadenas.

In [5]:
type('17')

str

In [6]:
type('3.2')

str

Son cadenas.

Cuando escribe un número entero grande, es posible que tenga la tentación de usar comas entre grupos de tres dígitos, como en 1,000,000. Este no es un número entero legal en Python, pero es legal:

In [7]:
print(1,000,00)

1 0 0


Bueno, eso no es lo que esperábamos en absoluto! Python interpreta 1,000,000 como una secuencia de enteros separados por comas, que imprime con espacios intermedios.


Este es el primer ejemplo que hemos visto de un error semántico: el código se ejecuta sin producir un mensaje de error, pero no hace lo "correcto".

## Variables

Una de las características más potentes de un lenguaje de programación es la capacidad de manipular variables. Una variable es un nombre que se refiere a un valor.

Una instrucción de asignación crea nuevas variables y le da valores:

In [8]:
mensaje = 'Y ahora algo completamente diferente'
n = 17
pi = 3.1415926535897931

Este ejemplo hace tres asignaciones. El primero asigna una cadena a una nueva variable llamada *mensaje*; el segundo asigna el entero 17 a *n*; el tercero asigna el valor (aproximado) de π a *pi*.

Para mostrar el valor de una variable, puede usar una sentencia *print*:

In [9]:
print(n)
print(pi)

17
3.141592653589793


El tipo de una variable es el tipo del valor al que se refiere.

In [10]:
type(mensaje)

str

In [11]:
type(n)

int

In [12]:
type(pi)

float

## Nombres de variables y palabras claves

Los programadores generalmente eligen nombres para sus variables que son significativos y documentan para qué se usa la variable.

Los nombres de variables pueden ser arbitrariamente largos. Pueden contener tanto letras como números, pero no pueden comenzar con un número. Es legal usar letras mayúsculas, pero es una buena idea comenzar los nombres de las variables con una letra minúscula (verás por qué más adelante).

El carácter de subrayado _ puede aparecer en un nombre. A menudo se usa en nombres con varias palabras, como mi_nombre o velocidad_de_vuelo. Los nombres de variables pueden comenzar con un carácter de guion bajo, pero generalmente evitamos hacerlo a menos que estemos escribiendo bibliotecas de código para que otros lo usen.

Si asigna a una variable un nombre ilegal, obtendrá un error de sintaxis:

In [13]:
76trombones = 'música'

SyntaxError: invalid syntax (<ipython-input-13-ee94c7e0b6d5>, line 1)

In [14]:
mor@ = 1000000

SyntaxError: invalid syntax (<ipython-input-14-1c9f450d7dd8>, line 1)

In [15]:
class = 'hola'

SyntaxError: invalid syntax (<ipython-input-15-5f8993154490>, line 1)

*76trombones* es ilegal porque comienza con un número. *more@* es ilegal, ya que contiene un carácter ilegal, *@*. Pero, ¿qué pasa con *class*?

Resulta que *class* una de las palabras clave de Python. El intérprete usa palabras clave para reconocer la estructura del programa, y no se pueden usar como nombres de variables.

Python reserva 33 palabras clave:

In [16]:
help("keywords")


Here is a list of the Python keywords.  Enter any keyword to get more help.

False               class               from                or
None                continue            global              pass
True                def                 if                  raise
and                 del                 import              return
as                  elif                in                  try
assert              else                is                  while
async               except              lambda              with
await               finally             nonlocal            yield
break               for                 not                 



Es posible que desee mantener esta lista a mano. Si el intérprete se queja de uno de sus nombres de variable y no sabe por qué, vea si está en esta lista.

## Declaraciones

Una declaración es una unidad de código que el intérprete de Python puede ejecutar. Hemos visto dos tipos de declaraciones: `print` siendo un enunciado y una asignación de expresión.

Cuando escribe una instrucción en modo interactivo, el intérprete la ejecuta y muestra el resultado, si hay alguno.

Un **script** generalmente contiene una secuencia de declaraciones. Si hay más de una declaración, los resultados aparecen de uno en uno a medida que se ejecutan las declaraciones.

Por ejemplo, la secuencia de comandos

In [17]:
print(1)
x = 2
print(x)

1
2


la instrucción de asignación no produce salida.

## Operadores y operandos

Los operadores son símbolos especiales que representan cálculos como la suma y la multiplicación. Los valores a los que se aplica el operador se llaman operandos.

Los operadores `+`, `-`, `*`, `/`, y `**` realizan la suma, resta, multiplicación, división y exponenciación, como en los siguientes ejemplos:

    20+32 hora-1 hora*60+minutos minutos/60 5**2 (5+9)*(15-7)

Ha habido un cambio en el operador de división entre Python 2.x y Python 3.x. En Python 3.x, el resultado de esta división es un resultado en coma flotante:

In [18]:
minuto = 59
minuto/60

0.9833333333333333

El operador de división en Python 2.0 dividiría dos enteros y truncaría el resultado en un entero.

Para obtener la misma respuesta en Python 3.0 use la división de piso (// entero)

In [19]:
minuto = 59
minuto//60

0

En Python 3.0 la división de enteros funciona mucho más de lo que cabría esperar si ingresa la expresión en una calculadora.


## Expresiones

Una expresión es una combinación de valores, variables y operadores. Un valor en sí mismo se considera una expresión, y también lo es una variable, por lo que las siguientes son todas expresiones legales (suponiendo que a la variable x se le ha asignado un valor):

In [20]:
17
x
x + 17

19

Si escribe una expresión en modo interactivo, el intérprete la evalúa y muestra el resultado:

In [21]:
1 + 1

2

¡Pero en un programa, una expresión por sí sola no hace nada! Esta es una fuente común de confusión para los principiantes.

Ejercicio 1: Escriba las siguientes declaraciones en el intérprete de Python para ver lo que hacen:

In [22]:
5
x = 5
x + 1

6

## Orden de operadores

Cuando aparece más de un operador en una expresión, el orden de evaluación depende de las reglas de precedencia. Para los operadores matemáticos, Python sigue la convención matemática. El acrónimo **PEMDAS** es una forma útil de recordar las reglas:
* Los **P**aréntesis tienen la precedencia más alta y se pueden usar para forzar que una expresión se evalúe en el orden que desee. Dado que las expresiones entre paréntesis se evalúan primero, 2 * (3 - 1) es 4 y (1 + 1)**(5 - 2) es 8. También puede usar paréntesis para hacer que una expresión sea más fácil de leer, como en (minuto * 100) / 60, incluso si no cambia el resultado.
* **E**xponenciación tiene la siguiente precedencia más alta, por lo que 2**1 + 1 es 3, no 4, y 3 * 1**3 es 3, no 27.
* **M**ultiplicación y **D**ivisión tienen la misma precedencia, que es más alta que **A**dición y **S**ustracción, que también tienen la misma precedencia. Entonces 2 * 3 – 1 es 5, no 4, y 6 + 4 / 2 es 8.0, no 5.
* Los operadores con la misma precedencia se evalúan de izquierda a derecha. Entonces la expresión 5 – 3 – 1 es 1, no 3, porque 5 – 3 sucede primero y luego 1 se resta de 2.

En caso de duda, siempre coloque paréntesis en sus expresiones para asegurarse de que los cálculos se realizan en el orden que desea.

## Operador de modulo

El operador de módulo trabaja en enteros y produce el resto cuando el primer operando se divide por el segundo. En Python, el operador de módulo es un signo de porcentaje (%). La sintaxis es la misma que para otros operadores:

In [23]:
cociente = 7 // 3
print(cociente)
resto = 7 % 3
print(resto)

2
1


Entonces 7 dividido por 3 es 2 con 1 sobrante.

El operador de módulo resulta ser sorprendentemente útil. Por ejemplo, puede verificar si un número es divisible por otro: si *x* % *y* es cero, entonces *x* es divisible por *y*.

También puede extraer el dígito o los dígitos del más a la derecha de un número. Por ejemplo, *x* % 10 arroja el dígito más a la derecha de *x* (en la base 10). Del mismo modo, *x* % 100 produce los dos últimos dígitos.

## Operadores de cadenas

El operador + trabaja con cadenas, pero no es una adición en el sentido matemático. En su lugar, realiza la concatenación, lo que significa unir las cadenas uniéndolas de extremo a extremo. Por ejemplo:

In [24]:
primero = 10
segundo = 15
print(primero + segundo)

25


In [25]:
primero = '100'
segundo = '150'
print(primero + segundo)

100150


El resultado de este ultimo programa es 100150.

## Pedirle al usuario su opinión

A veces nos gustaría tomar el valor de una variable del usuario a través de su teclado. Python proporciona una función incorporada llamada *input* que obtiene la entrada desde el teclado. Cuando se llama a esta función, el programa se detiene y espera a que el usuario escriba algo. Cuando el usuario presiona Return o Enter, el programa se reanuda y *input* devuelve en forma de cadena lo que el usuario tipeó.

In [26]:
entrada = input()
print(entrada)

Hola!
Hola!


Antes de recibir información del usuario, es una buena idea imprimir un mensaje que le indique al usuario qué debe ingresar. Puede pasar una cadena a *input* para que se muestre al usuario antes de hacer una pausa para la entrada:

In [27]:
nombre = input('Cuál es tu nombre?\n')
print(nombre)

Cuál es tu nombre?
Chuck
Chuck


La secuencia *\n* al final del prompt representa una nueva línea, que es un carácter especial que causa un salto de línea. Es por eso que la entrada del usuario aparece debajo del aviso.

Si espera que el usuario escriba un número entero, puede intentar convertir el valor de retorno a *int* con la función `int()`:

In [28]:
prompt = 'Cuál es la velocidad de una golondrina?\n'
velocidad = input(prompt)
int(velocidad)
int(velocidad) + 5

Cuál es la velocidad de una golondrina?
75


80

In [29]:
prompt = 'Cuál es la velocidad de una golondrina?\n'
velocidad = input(prompt)
int(velocidad)
int(velocidad) + 5

Cuál es la velocidad de una golondrina?
75


80

Veremos cómo manejar este tipo de error más adelante.

## Comentarios

A medida que los programas crecen y se vuelven más complicados, se vuelven más difíciles de leer. Los lenguajes formales son densos, y a menudo es difícil mirar un fragmento de código y descubrir qué está haciendo o por qué.

Por esta razón, es una buena idea agregar notas a sus programas para explicar en lenguaje natural lo que está haciendo el programa. Estas notas se llaman comentarios, y en Python comienzan con el símbolo #:

In [30]:
# calcular el porcentaje de horas que han transcurrido
porcentaje = (minuto * 100) / 60

En este caso, el comentario aparece solo en una línea. También puedes poner comentarios al final de una línea:

In [31]:
porcentaje = (minuto * 100) / 60 # porcentaje de una hora

Todo lo que esta después de # hasta el final de la línea se ignora; no tiene ningún efecto en el programa.

Los comentarios son más útiles cuando documentan características no obvias del código. Es razonable suponer que el lector puede descubrir qué hace el código; es mucho más útil explicar por qué.

Este comentario es redundante con el código e inútil:

In [32]:
v = 5 # asigna 5 a v

Este comentario contiene información útil que no está en el código:

In [33]:
v = 5 # velocidad en metros/segundos.

Los buenos nombres de variables pueden reducir la necesidad de comentarios, pero los nombres largos pueden hacer que las expresiones complejas sean difíciles de leer, por lo que hay una solución de compromiso.

## Elegir nombres de variables mnemotécnicas

Siempre que siga las reglas simples de nomenclatura en las variables y evite las palabras reservadas, tendrá muchas opciones cuando nombre sus variables. Al principio, esta elección puede ser confusa tanto cuando lee un programa como cuando escribe sus propios programas. Por ejemplo, los siguientes tres programas son idénticos en términos de lo que logran, pero muy diferentes cuando los lee y trata de comprenderlos.

In [34]:
a = 35.0
b = 12.50
c = a * b
print(c)

437.5


In [35]:
horas = 35.0
tarifa = 12.50
pago = horas * tarifa
print(pago)

437.5


In [36]:
x1q3z9ahd = 35.0
x1q3z9afd = 12.50
x1q3p9afd = x1q3z9ahd * x1q3z9afd
print(x1q3p9afd)

437.5


El intérprete de Python considera que estos tres programas son exactamente iguales, pero los humanos ven y entienden estos programas de manera bastante diferente. Los humanos comprenderán más rápidamente la intención del segundo programa porque el programador ha elegido nombres de variables que reflejan su intención con respecto a qué datos se almacenarán en cada variable.

Llamamos a estos nombres de variables sabiamente elegidos "nombres de variables mnemotécnicas". La palabra mnemónico significa "ayuda con la memoria". Elegimos nombres de variables mnemotécnicas para ayudarnos a recordar por qué creamos la variable en primer lugar.

Aunque todo esto suena muy bien, y es una muy buena idea usar nombres de variables mnemotécnicos, los nombres de variables mnemotécnicos pueden interferir en la capacidad de un programador principiante de analizar y comprender el código. Esto se debe a que los programadores principiantes aún no han memorizado las palabras reservadas (solo hay 33) y algunas veces las variables con nombres que son demasiado descriptivos comienzan a parecerse a una parte del lenguaje y no solo a nombres de variables bien elegidos.

Eche un vistazo rápido al siguiente código de ejemplo de Python que recorre algunos datos. Pronto cubriremos los bucles, pero por ahora trataremos de descifrar lo que esto significa:

In [37]:
palabras = ['Yo','voy','a','aprender','Python']
for palabra in palabras:
    print(palabra)

Yo
voy
a
aprender
Python


¿Qué está sucediendo aquí? ¿Cuáles de los tokens (*for*, palabra, *in*, etc.) son palabras reservadas y cuáles son solo nombres de variables? ¿Python entiende en un nivel fundamental la noción de palabras? Los programadores principiantes tienen problemas para separar qué partes del código deben ser iguales a este ejemplo y qué partes del código son simplemente elecciones hechas por el programador.

El siguiente código es equivalente al código anterior:

In [38]:
pizza = ['rebanada1','rebanada2','rebanada3','rebanada4']
for rebanada in pizza:
    print(rebanada)

rebanada1
rebanada2
rebanada3
rebanada4


Es más fácil para el programador principiante mirar este código y saber qué partes son palabras reservadas definidas por Python y qué partes son simplemente nombres variables elegidos por el programador. Es bastante claro que Python no tiene una comprensión fundamental de la pizza y las rodajas y el hecho de que una pizza consiste en un conjunto de una o más rebanadas.

Pero si nuestro programa es realmente acerca de leer datos y buscar palabras en los datos, *pizza* y *rebanada* son nombres de variables muy poco nemotécnicos. Elegirlos como nombres variables distrae del significado del programa.

Después de un período de tiempo bastante corto, sabrá las palabras reservadas más comunes y comenzará a ver las palabras reservadas saltando hacia usted:

<pre><code><b>for</b> palabra <b>in</b> palabras<b>:</b>
    <b>print</b>(palabra)
</code></pre>

Las partes del código que se definen por Python (for, in, print, y :) están en negrita y las variables elegidas por el programador (*palabra* y *palabras*) no están en negrita. Muchos editores de texto conocen la sintaxis de Python y colorearán las palabras reservadas de manera diferente para darte pistas para mantener separadas tus variables y palabras reservadas. Después de un tiempo, comenzará a leer Python y determinar rápidamente qué es una variable y qué es una palabra reservada.

## Depuración

En este punto, el error de sintaxis que tienen más probabilidades de hacer es un nombre de variable ilegal, como *class* y *yield*, que son palabras clave, o odd~job y US$, que contienen caracteres no válidos.
Si pone un espacio en un nombre de variable, Python piensa que son dos operandos sin un operador:

In [39]:
mal nombre = 5

SyntaxError: invalid syntax (<ipython-input-39-44564c6bebed>, line 1)

In [40]:
mes = 09

SyntaxError: invalid token (<ipython-input-40-a9ab09a355cb>, line 1)

Para los errores de sintaxis, los mensajes de error no ayudan mucho. Los mensajes más comunes son

    SyntaxError: invalid syntax 

y 

    SyntaxError: invalid token

ninguno de los cuales es muy informativo.

El error de tiempo de ejecución que es más probable que suceda es un "uso antes de la definición"; es decir, tratando de usar una variable antes de haber asignado un valor. Esto puede suceder si escribes mal el nombre de una variable:

In [41]:
principal = 327.68
intereses = principl * tarifa

NameError: name 'principl' is not defined

Los nombres de las variables son sensibles a las mayúsculas y minúsculas, por lo que LaTeX no es lo mismo que latex.

En este punto, la causa más probable de un error semántico es el orden de las operaciones. Por ejemplo, para evaluar de 1/2 π , que podría estar tentado a escribir

In [42]:
1.0 / 2.0 * pi

1.5707963267948966

Pero la división ocurre primero, por lo que obtendrías π / 2 , ¡que no es lo mismo! No hay forma de que Python sepa lo que quería escribir, por lo que en este caso no obtendrá un mensaje de error; simplemente recibes la respuesta incorrecta

## Glosario

* **asignación:** Una declaración que asigna un valor a una variable.
* **concatenar:** Para unir dos operandos de extremo a extremo.
* **comentario:** Información en un programa que está destinado a otros programadores (o cualquier persona que lea el código fuente) y no tiene ningún efecto en la ejecución del programa.
* **evaluar:** Simplificar una expresión realizando las operaciones para producir un solo valor.
* **expresión:** Una combinación de variables, operadores y valores que representa un único valor de resultado.
* **punto flotante (float):** Un tipo de dato que representa números con partes fraccionales.
* **entero (int):** Un tipo de dato que representa números enteros.
* **palabra clave:** Una palabra reservada que el compilador usa para analizar un programa; no se puede utilizar como palabras clave if, def y while como nombres de variables.
* **mnemotécnico:** Una ayuda de memoria. A menudo damos nombres mnemotécnicos de variables para ayudarnos a recordar qué se almacena en la variable.
* **operador de módulo:** Un operador denotado con un signo de porcentaje (%), que trabaja en enteros y cede el resto cuando un número se divide por otro.
* **operando:** Uno de los valores en los que opera un operador.
* **operador:** Un símbolo especial que representa un cálculo simple como adición, multiplicación o
concatenación de cadenas.
* **reglas de precedencia:** El conjunto de reglas que rige el orden en que se evalúan las expresiones que involucran operadores múltiples y operandos.
* **declaración:** Una sección de código que representa un comando o acción. Hasta el momento, las declaraciones que hemos visto son asignaciones y declaración de expresión de impresión.
* **string (str):** Un tipo de dato que representa secuencias de caracteres.
* **tipo:** Una categoría de valores. Los tipos de datos que hemos visto hasta ahora son enteros (tipo int), números de coma flotante (tipo float) y cadenas (tipo str).
* **valor:** Una de las unidades básicas de datos, como un número o cadena, que un programa manipula.
* **variable:** Un nombre que se refiere a un valor.

## Ejercicios

**Ejercicio 1:**  escriba un programa que utilice *input* para solicitarle un nombre a un usuario y luego le da la bienvenida.
    
    Ingresa tu nombre: Chuck
    Hola Chuck
    
**Ejercicio 2:** escriba un programa para solicitar al usuario las horas y tarifa por hora para calcular el salario bruto.

    Ingresa las horas: 35
    Ingresa la tarifa: 2.75
    Salario bruto: 96.25
    
No nos preocuparemos por asegurarnos de que nuestro pago tenga exactamente dos dígitos después del decimal por ahora. Si lo desea, puede jugar con la función incorporada de Python llamada *round* para redondear adecuadamente el pago resultante a dos decimales.

**Ejercicio 3:**  suponga que ejecutamos las siguientes declaraciones de asignación:

    ancho = 17
    alto = 12.0
    
Para cada una de las siguientes expresiones, escriba el valor de la expresión y el tipo (del valor de la expresión).

1. ancho//2
2. ancho/2.0
3. alto/3
4. 1 + 2

Use el intérprete de Python para verificar sus respuestas.

**Ejercicio 4:** escriba un programa que le solicite al usuario una temperatura en Celsius, y la convierta a Fahrenheit e imprima la temperatura convertida.

<!--NAVIGATION-->
| [Indice](indice.ipynb) | 

< [Capítulo 1 - ¿Por qué deberías aprender a escribir código?](cap01.ipynb) | [Capítulo 3 - Ejecuciones condicionales](cap03.ipynb) >